canvas_sdk/data/base.py

```diff
@@ -0,0 +1,15 @@
+import json
+
+from pydantic import BaseModel, ConfigDict
+
+from canvas_sdk.utils.json import JSONEncoder
+
+
+class PluginModel(BaseModel):
+    model_config = ConfigDict(strict=True, validate_assignment=True)
+
+    def as_json(self, nested_data_dict=True, exclude_unset=False) -> str:
+        json_dict = self.model_dump(exclude_unset=exclude_unset)
+        if nested_data_dict is True:
+            json_dict = {"data": json_dict}
+        return json.dumps(json_dict, cls=JSONEncoder)
```

Comment:
I'm going to propose a slightly different design here.

First suggestion: We can specify the JSON encoding options in the model config:

```python
json_encoders = {
    date: lambda v: v.isoformat(),
    Enum: lambda v: v.value,
}
```

This has the benefit that callers of the base class's `json` method will get a correctly-encoded result.

Second suggestion: Just reusing `json` (rather than having a separate method) would be ideal for cases where we don't want nested data.

Third suggestion: Rather than a boolean, just make a dedicated method for the nested case. Additionally, to avoid having to recurse through the model twice, we could just do some kludgy string concatenation:

```python
def json_nested(self, **args, **kwargs) -> str:
    return f"\{\"data\":{self.json(*args, **kwargs)}\}"
```

Alternatively, we could override `json` and add a kwarg for the nested argument. Some feel that boolean parameters are an antipattern, but the `json` method already has boolean parameters, so 🤷

Reply:
I like this approach. It looks like `json_encoders` is removed in pydantic v2, which we are using - https://github.com/pydantic/pydantic/discussions/6697.

I'll take a look at the alternative best practice for this.

Reply:
According to the documentation, `json_encoders` is deprecated, but they left it in because they don't have a direct replacement yet:

https://docs.pydantic.dev/latest/api/config/#pydantic.config.ConfigDict.json_encoders

I read that as we can/should continue using it until they do provide a direct replacement.

Reply:
@csande good find. I'll use it. 👍

Reply:
It also looks like the `json` method is deprecated, and that `model_dump_json` is the replacement.

Reply:
I just pushed changes to add `json_encoders`. I also made a new method called `model_dump_json_nested`, which I think is cleaner than overriding and adding an argument to `model_dump_json`.